### DISCUSSION TITLE
Extreme ERS = Avaya ERS backup Start & Running config

### DISCUSSION THREAD

**Discussion by Microbuse** [ktbyers/netmiko#2374] [General]
Hi
Do anyone have working script with netmiko to backup Extreme ERS.
I cant get  wait to work. I have maid two scipts. One with telnetlib and one on ssh paramiko but it is not reliable. it users
I can´t get it to wait on backup to finish. If I use time.sleep it works. Stacks take longer time 1024bytes per unit and then all the odder switches also will have the high sleep time. That will make the script run 24/7 

//Micke

**Comment by Microbuse** [comment]
[Backup_Multi_Avaya_ERS.txt](https://github.com/ktbyers/netmiko/files/6830995/Backup_Multi_Avaya_ERS.txt)

**Comment by Microbuse** [comment]
Works now with time.sleep=10 but want it to sense when the file is done.
Also "device_type": "avaya_ers" does not work with Avaya Wireless LAN Controller WC8180 that have Meny.
 enter = "\r\n"      # Enter "New Line"
    command0 = "\x19"   # Enter Ctrl-Y
    command1 = user     # Enter username
    command2 = password # Enter password
    command3 = "\x63"   # Get to command line
    command4 = "en"     # Get to enable mode
    command5 = "copy running-config tftp address %s filename %s_Run_%s\r" % (tftp_server, ip_address, str(datetime.now().strftime("%d%m%Y_%H%M")))
    command6 = "copy config tftp address %s filename %s_Str_%s\r" % (tftp_server, ip_address, str(datetime.now().strftime("%d%m%Y_%H%M")))
    command7 = "exit" # Logout from cli to menu for ERS
    command8 = "\x6C" # Logout session
    command9 = "\x71" # Logout session any key = q

**Comment by ktbyers** [comment]
@Microbuse Can you show what a manual CLI session of doing this looks like? That will help me understand what needs done when using Netmiko.

**Reply by Microbuse** [reply]
Sorry bean on vacation but now I am back. Added some pic

**Comment by Microbuse** [comment]
![Ctrl-Y](https://user-images.githubusercontent.com/81070908/126214505-574b2676-7fa2-47c0-9d0c-0a5ba9255cb0.JPG)
![Meny](https://user-images.githubusercontent.com/81070908/126214525-078f4077-1bf2-4b5b-84b7-6092ac70ca5d.JPG)

**Reply by Microbuse** [reply]
You connect to ip and login id/password, then hit Ctrl-Y "\x19" like the odder switches. but then you get to a Menu ware you need to run Command Line "\x63"
Also when exiting this switch you get to Menu on Command Line "exit" ware you need to Logout "\x6C"

Let me know if you need more info.

//Micke

**Reply by ktbyers** [reply]
You will need a special login handler look at the Calix B6 driver in Netmiko for an example. Someone will have to create this for the given device_type (or create a new device_type for it).

I would also see if you can completely disable this behavior via configuration on the device.

**Reply by kadecole** [reply]
I know on Avaya/Nortel devices you can issue the `cmd-interface cli` command and when you next login you will be put into the cli and not the menu. I do not know if this works on the Wireless Lan controllers.

**Reply by Microbuse** [reply]
Hi Did an ugly fix. but works for now.

**Reply by Microbuse** [reply]
![multi](https://user-images.githubusercontent.com/81070908/129678651-55e2151b-fa2a-4ed0-b39e-17fee9a88322.JPG)

**Reply by romanstech** [reply]
@Microbuse @ktbyers Based on Microbuse's picture, I've finally succeeded to connect to Avaya ERS without `cmd-interface cli` command. 
See below a couple of rows that I've added/edited to file **/lib/python3.8/site-packages/netmiko/extreme/extreme_ers_ssh.py**. Netmiko v.4.1.1.
```

CTRL_Y = "\x19"
CTRL_C = "\x63"
...
cntl_y = "Ctrl-Y"
menu = "Menu"
pattern = rf"(?:{uname}|{password}|{cntl_y}|{self.prompt_pattern}|{self.enable_pattern}|{menu})"
...
if cntl_y in new_data:
    self.write_channel(CTRL_Y)
elif menu in new_data:
    self.write_channel(CTRL_C)
```

I've checked this solution on 5698, 5520 and 5510 switches. All with Menu but without Ctrl-Y that I've disabled a long time ago with `banner disabled` or `no banner` depending on switch type and ERS version.

```
# diff .venv/lib/python3.8/site-packages/netmiko/extreme/extreme_ers_ssh.py .venv/lib/python3.8/site-packages/netmiko/extreme/extreme_ers_ssh.py_orig
1c1
< """Netmiko support for Extreme/Avaya Ethernet Routing Switch."""
---
> """Netmiko support for Extreme Ethernet Routing Switch."""
6c6
< # Extreme ERS presents Enter Ctrl-Y to begin and Ctrl-C to pass start-up menu.
---
> # Extreme ERS presents Enter Ctrl-Y to begin.
8d7
< CTRL_C = "\x63"
15d13
<     enable_pattern = "#"
25c23
<         Extreme/Avaya ERS presents the following as part of the login process:
---
>         Extreme ERS presents the following as part of the login process:
27c25
<         Enter Ctrl-Y to begin (can be disabled with banner disabled or no banner config command)
---
>         Enter Ctrl-Y to begin
32,40d29
<
<         Then we will see Menu of commands which can be disabled with cmd-interface cli config command
<
<         But now this script allows to connect SSH to devices with Ctrl-Y and Menu presented.
<
<         Tested with ERS v.6.3 and higher on 5510, 5510, 5698 Avaya ERS switches.
<
<         Used "device_type": "avaya_ers"
<
47,48c36
<         menu = "Menu"
<         pattern = rf"(?:{uname}|{password}|{cntl_y}|{self.prompt_pattern}|{self.enable_pattern}|{menu})"
---
>         pattern = rf"(?:{uname}|{password}|{cntl_y}|{self.prompt_pattern})"
57,58d44
<             elif menu in new_data:
<                 self.write_channel(CTRL_C)
```

**Comment by Microbuse** [comment]
Hi

Thank. I will look in to this.
This is only temp backup fix  until we move to arista.

//Micke

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Reply by ktbyers** [reply]
@Microbuse You might want to submit that as a PR so other people can leverage that as well.

**Reply by Microbuse** [reply]
Hi. I am new to this but I installed the Desktop version and open the netmiko repository. Do I then create a new branch and publish the changed code? name example extreme_ers_ssh_with_menu

**Reply by Microbuse** [reply]
@ktbyers I think I managed to do an PR.
